gems/pending/openstack/openstack_handle/handled_list.rb

```diff
@@ -0,0 +1,118 @@
+        @os_handle.accessor_for_accessible_tenants(self.class::SERVICE_NAME,
+                                                   paginated_list_method,
+                                                   identity_method(collection_type))
+      when :all_tenants_option
+        get_paginated_list(collection_type, options.merge(:all_tenants => 'True'), method)
+      when :none
+        get_paginated_list(collection_type, options, method)
+      else
+        # Raise error about unknown multitenancy type and return list for current tenant only
+        raise "Unknown multitenancy type '#{multi_tenancy_type(collection_type)}' for collection #{collection_type}"
+      end
+    end
+
+    def get_paginated_list(collection_type, options = {}, method = :all)
+      case pagination_type(collection_type)
+      when :marker
+        get_marker_based_paginated_list(collection_type, options, method)
+      when :page_number
+        get_page_number_based_paginated_list(collection_type, options, method)
+      when :none
+        call_list_method(collection_type, options, method)
+      else
+        # Raise error about unknown pagination type and return non paginated list
+        raise "Unknown pagination type '#{pagination_type(collection_type)}' for collection #{collection_type}"
+      end
+    rescue Excon::Errors::Forbidden => err
+      # It can happen user doesn't have rights to read some tenant, in that case log warning but continue refresh
+      _log.warn "Forbidden to read the tenant, Message=#{err.message}"
+      _log.warn err.backtrace.join("\n")
+      []
+    end
+
+    private
+
+    def get_marker_based_paginated_list(collection_type, options, method)
+      all_servers = call_list_method(collection_type, options, method)
+      last_server = all_servers.try(:last)
+
+      while marker(last_server, collection_type)
+        servers_on_page = call_list_method(collection_type,
+                                           options,
+                                           method,
+                                           :marker => marker(last_server, collection_type),
+                                           :limit  => default_pagination_limit)
+        last_server = servers_on_page.try(:last)
+        break if pagination_break_condition(all_servers, servers_on_page, collection_type)
+        all_servers.concat(servers_on_page) unless servers_on_page.blank?
+      end
+
+      all_servers
+    end
+
+    def get_page_number_based_paginated_list(_collection_type, _options, _method)
+      # TBD, used e.g. by keystone v3
+    end
+
+    def call_list_method(collection_type, options, method, paginate_options = {:limit => default_pagination_limit})
+      public_send(collection_type).public_send(method, options.merge(paginate_options))
+    end
+
+    def marker(last_server, collection_type)
+      last_server.send(identity_method(collection_type)) if last_server && identity_method(collection_type)
```

Comment:
`&& identity_method(collection_type)` sounds a lot like apocalypse protection... if that's ever false, I'd much rather we just let it fail